lib/seattleflu/db/cli/command/etl/presence_absence.py

```diff
@@ -0,0 +1,397 @@
+    Select presence_absence test target by *identifier*, or insert it if it doesn't exist.
+    """
+    LOG.debug(f"Looking up target «{identifier}»")
+
+    target = db.fetch_row("""
+        select target_id as id, identifier
+          from warehouse.target
+         where identifier = %s
+        """, (identifier,))
+
+    if target:
+        LOG.info(f"Found target {target.id} «{target.identifier}»")
+    else:
+        LOG.debug(f"Target «{identifier}» not found, adding")
+
+        data = {
+            "identifier": identifier,
+            "control_status": control_status
+        }
+
+        target = db.fetch_row("""
+            insert into warehouse.target (identifier, control_status)
+                values (%(identifier)s, %(control_status)s)
+            returning target_id as id, identifier
+            """, data)
+
+        LOG.info(f"Created target {target.id} «{target.identifier}»")
+
+    return target
+
+
+def target_control_status(control_status: str, target_identifier: str) -> bool:
+    """
+    Determine the control status of the target.
+    """
+    expected_values = ["NotControl", "PositiveControl"]
+    if not control_status or control_status not in expected_values: 
+        raise UnknownControlStatusError(f"Unknown control status «{control_status}» for target {target_identifier}")
+    return control_status == "PositiveControl"
+
+
+def find_or_create_sample(db: DatabaseSession, identifier: str,
+                          details: dict) -> Any:
+    """
+    Select sample by *identifier*, or insert it if it doesn't exist.
+
+    TODO eventually we'll need to add the encounter_id once the relationship
+    between encounter_id and sample_id exists in the db.
+    """
+    LOG.debug(f"Looking up sample «{identifier}»")
+
+    sample = db.fetch_row("""
+        select sample_id as id, identifier
+          from warehouse.sample
+         where identifier = %s
+        """, (identifier,))
+
+    if sample:
+        LOG.info(f"Found sample {sample.id} «{sample.identifier}»")
+    else:
+        LOG.debug(f"Sample «{identifier}» not found, adding")
+
+        data = {
+            "identifier": identifier,
+            "details": Json(details)
+            # "encounter_id"
+        }
+
+        sample = db.fetch_row("""
+            insert into warehouse.sample (
+                identifier, 
+                details)
+                
+                values (
+                    %(identifier)s, 
+                    %(details)s)
+                    
+            returning sample_id as id, identifier
+            """, data)
+
+        LOG.info(f"Created sample {sample.id} «{sample.identifier}»")
+
+    return sample
+
+def sample_identifier(db: DatabaseSession, barcode: str) -> str:
+    """
+    Find corresponding UUID for scanned sample barcode within
+    warehouse.identifier.
+
+    Returns barcode if it starts with "SCH-" as this indicates it 
+    is a control sample, and therefore does not have a UUID.
```

Comment:
This bit of doc appears to be out of date? Nothing tests for `barcode.startswith("SCH-")`...